server/sonar-server/src/main/java/org/sonar/server/computation/step/NewCoverageMeasuresStep.java

```diff
@@ -0,0 +1,353 @@
+
+  public static final class Counter {
+    private final int index;
+    private final Date date;
+    private Integer newLines;
+    private Integer newCoveredLines;
+    private Integer newConditions;
+    private Integer newCoveredConditions;
+
+    private Counter(int index, @Nullable Date date) {
+      this.index = index;
+      this.date = date;
+    }
+
+    public int getIndex() {
+      return index;
+    }
+
+    void analyze(@Nullable Date lineDate, int hits, int conditions, int coveredConditions) {
+      if (lineDate == null) {
+        // TODO warning
+
+      } else if (date == null || lineDate.after(date)) {
+        // TODO test if string comparison is faster or not
+        addLine(hits > 0);
+        addConditions(conditions, coveredConditions);
+      }
+    }
+
+    void addLine(boolean covered) {
+      if (newLines == null) {
+        newLines = 0;
+      }
+      newLines += 1;
+      if (covered) {
+        if (newCoveredLines == null) {
+          newCoveredLines = 0;
+        }
+        newCoveredLines += 1;
+      }
+    }
+
+    void addConditions(int count, int countCovered) {
+      if (newConditions == null) {
+        newConditions = 0;
+      }
+      newConditions += count;
+      if (count > 0) {
+        if (newCoveredConditions == null) {
+          newCoveredConditions = 0;
+        }
+        newCoveredConditions += countCovered;
+      }
+    }
+
+    boolean hasNewCode() {
+      return newLines != null;
+    }
+
+    public int getNewLines() {
+      return newLines != null ? newLines : 0;
+    }
+
+    public int getNewCoveredLines() {
+      return newCoveredLines != null ? newCoveredLines : 0;
+    }
+
+    public int getNewConditions() {
+      return newConditions != null ? newConditions : 0;
+    }
+
+    public int getNewCoveredConditions() {
+      return newCoveredConditions != null ? newCoveredConditions : 0;
+    }
+  }
+
+  private enum PeriodToCounter implements Function<Period, Counter> {
+    INSTANCE;
+
+    @Override
+    @Nonnull
+    public Counter apply(@Nonnull Period input) {
+      return new Counter(input.getIndex(), new Date(input.getSnapshotDate()));
+    }
+  }
+
+  private enum CounterHasNewCode implements Predicate<Counter> {
+    INSTANCE;
+
+    @Override
+    public boolean apply(@Nonnull Counter input) {
+      return input.hasNewCode();
+    }
+  }
+
+  /**
+   * Represents a way of computing a measure value from a given Counter.
+   */
+  private interface NewCoverageMeasureComputer {
+    double compute(Counter counter);
+  }
+
+  private enum NewLinesToCoverComputer implements NewCoverageMeasureComputer {
+    INSTANCE;
+
+    @Override
+    public double compute(Counter counter) {
+      return (double) counter.getNewLines();
+    }
+  }
+
+  private enum NewUncoveredLinesComputer implements NewCoverageMeasureComputer {
+    INSTANCE;
+
+    @Override
+    public double compute(Counter counter) {
+      return counter.getNewLines() - counter.getNewCoveredLines();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Cast one of the operands of this substraction operation to a "double". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2184)